### PR TITLE
Throw on multiple property definitions for the same node

### PR DESCRIPTION
**Pull request by FlorentinD**
Based on #68.

I didn't see a good way to check for duplicate vertex definitions.
However I could find a way to throw, when properties are assigned to an existing vertex.
Previously the second property assignment was just silently ignored.

Other options would be to allow these later property assignments, when there was none in the first declaration like `(v1), (v1 {name:Bob}` or just throwing on duplicate keys.

### CODE REVIEW

File: src/main/java/org/s1ck/gdl/GDLLoader.java [hunk 1]
```diff
@@ -338,6 +338,9 @@ public void enterVertex(GDLParser.VertexContext vertexContext) {
     Vertex v;
     if (variable != null && userVertexCache.containsKey(variable)) {
       v = userVertexCache.get(variable);
+      if (!getProperties(vertexContext.properties()).isEmpty()) {
```
Comment (s1ck) [inline src/main/java/org/s1ck/gdl/GDLLoader.java:341]: We should also check if the vertex context has label definitions. Maybe it makes sense to have sth like:
```
private boolean isEmpty(GDLParser.VertexContext vertexContext) {
    return vertexContext.header().label().isEmpty() && vertexContext.properties().property().isEmpty();
}
```

File: src/main/java/org/s1ck/gdl/GDLLoader.java [hunk 2]
```diff
@@ -338,6 +338,9 @@ public void enterVertex(GDLParser.VertexContext vertexContext) {
...
+        throw new IllegalArgumentException("Properties for vertex " + variable + " defined multiple times");
```
Comment (s1ck) [inline src/main/java/org/s1ck/gdl/GDLLoader.java:342]: I think the text could be something like:
```
String.format("Vertex `%1$s` is declared multiple times. For reference, consider using `(%1$s)`.", variable)
```

**Review by s1ck — CHANGES_REQUESTED** [review]
I like idea, although it's a breaking change. However, since graphs and relationships behave in a similar way, the same check should be applied here, too.

File: src/main/java/org/s1ck/gdl/GDLLoader.java (hunk 1 above)
Comment (FlorentinD) [reply src/main/java/org/s1ck/gdl/GDLLoader.java:341]: I agree, making it clear, that any vertex detail should be set in its first declartion

File: src/main/java/org/s1ck/gdl/GDLLoader.java (hunk 2 above)
Comment (FlorentinD) [reply src/main/java/org/s1ck/gdl/GDLLoader.java]: I chose to replace the second part with a more general sentence regarding entities.

File: src/main/java/org/s1ck/gdl/GDLLoader.java [hunk 3]
```diff
@@ -304,6 +308,12 @@ public void enterGraph(GDLParser.GraphContext graphContext) {
     currentGraphId = g.getId();
   }
 
+  private boolean isEmpty(GDLParser.GraphContext ctx) {
```
Comment (s1ck) [inline src/main/java/org/s1ck/gdl/GDLLoader.java:311]: Maybe we can just have `isEmpty(HeaderContext header, PropertiesContext  properties)` and use that function in all three places?

File: src/main/java/org/s1ck/gdl/exceptions/DuplicateDeclarationException.java [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+ */
+public class DuplicateDeclarationException extends RuntimeException {
+
+  /**
+   * Creates a new exception
+   *
+   * @param variable the variable which could not be resolved
+   */
+  public DuplicateDeclarationException(String variable, String entity) {
+    super(String.format("%1$s `%2$s` is declared multiple times. Do not declare properties or labels while referencing a variable.", entity, variable));
```
Comment (s1ck) [inline src/main/java/org/s1ck/gdl/exceptions/DuplicateDeclarationException.java:30]: I don't think you need to use the indexed arguments here. It's mostly useful if you reuse the same argument multiple times.

Comment (s1ck) [inline src/main/java/org/s1ck/gdl/exceptions/DuplicateDeclarationException.java:30]:
```suggestion
    super(String.format("%1$s `%2$s` is declared multiple times. Declaring properties or labels while referencing a variable is not allowed. Use `(%2$s)` to refer to the element instead.", entity, variable));
```

**Review by s1ck — APPROVED** [review]

**Pull request merged**